### ISSUE TITLE
Italia Irredenta spammy bug

### ISSUE THREAD

**Issue by rogerburks** [moretrim/PFH#39]
The decision italia_irredenta in Italy.txt can become allowed but not satisfied by the effect, such that it becomes spammy when the ai runs Italy, until Italy finally achieves the situation that allows the effect to be implemented. This likely has several deleterious effects. I have attempted a fix in this, but it is a very complex decision and my fix may have disabled it. More testing can reveal whether this is the case or not.

**Comment by rogerburks** [comment]
Here's a summary of this. In the original, the `allow` statement is highly permissive relative to the considerably more complex `effect` statement, and therefore the ai simply spams this decision until the effect can do something:

https://github.com/moretrim/PFH/blob/bc42da48dbfbc09d2257055a9c8a839b33a111e7/PFH/decisions/Italy.txt#L673-L683

This seems to be work this way because each effect requires a different prerequisite tech and different miscellaneous conditions. For instance, here is the `effect` statement for Savoy, which is entirely different from the effects for other provinces affected by this decision:

https://github.com/moretrim/PFH/blob/bc42da48dbfbc09d2257055a9c8a839b33a111e7/PFH/decisions/Italy.txt#L724-L748

My approach to solving this has been to hopefully equalize each `allow` with the conditions requested by its `effect` while still allowing the decision to occur in the complex manner that seems to be intended. I am doing this by converting the allow to a more complex `OR` statement that is essentially copied from the effects.

The first attempt with Aosta seems to have worked, and the next attempts will be for all the other provinces affected by the decision.

**Comment by rogerburks** [comment]
As I could have guessed previously, `random_owned` breaks `allow` statements and prevents any `effect` from firing. However, `THIS` works just fine. Therefore the decision is hopefully fixed by converting an effect such as Savoia's above into an `allow` like this (with the old allow preserved after the `THIS` statement):

https://github.com/moretrim/PFH/blob/1d9dab4261b2634b70ec328848b79b08c5b8805f/PFH/decisions/Italy.txt#L704-L723

Not sure if it is fully working now, but at least it gave an Italy core to Venice and stopped being spammy on the latest playthrough.

**Comment by moretrim** [comment]
> As I could have guessed previously, `random_owned` breaks `allow` statements and prevents any `effect` from firing.

`random_owned` is an `effects` scope and is not allowed in triggers. In some sense `any_owned_province` is the closest trigger equivalent, *however* many uses of `random_owned` are really for building a conditional where the trigger in the `limit = { … }` block guards the follow-up effects.

In the latter case what really matters is what's inside the `limit` block, and what should appear in the decision trigger.

**Comment by rogerburks** [comment]
Update: Although I had initially wondered if this fix was working, it seems to now be doing fine. The decision is not spammy in my playthroughs anymore, and a reasonably successful Italy acquires its cores appropriately:

![annecyHasCores](https://user-images.githubusercontent.com/17787203/69492163-6e127800-0e53-11ea-82e8-db04924e5230.png)

![triestHasCores](https://user-images.githubusercontent.com/17787203/69492164-78cd0d00-0e53-11ea-8331-753964b88ccb.png)

**Comment by rogerburks** [comment]
Here it seemed like a good idea, after discussion, to examine the flags and other statements involved in this relatively large decision.

In the original decision's `potential`:
`italian_claimed_south_tirol` - this flag is not set anywhere or referenced anywhere else. It seems that this flag is not really needed, and it may have been an early attempt in default HFM to fix the spammy problem. Here I propose deleting this flag's reference. Part of my reasoning here is that the flag could get set even when the effect does set the intended cores.

`italian_reclaim_aosta` - as above, this flag is not set anywhere or referenced anywhere else. Again I propose deleting this reference.

`italian_reclaim_savoia` - again, a flag that is referenced but never set or referenced anywhere else.

`plombieres` - this flag is set in the `plombieres` decision in `Italy.txt`. It seems to be used, and so I plan to leave it alone.

In the original decision's `effect`:
`italian_claimed_venetia` - this flag is used in the `italian_islands` decision, and again in this decision before Trieste and South Tirol can be claimed. This looks like a flag that should remain unaltered.

`the_treaty_of_milan` - this flag is set in the `FlavourMod_FRA.txt` file. Again, should be kept.

So in summary, I suggest getting rid of the unused flags, because this decision's `effect` contains many conditionals, and because the flags do not seem to be needed.

**Comment by rogerburks** [comment]
Here I plan to modernize this issue a bit, and prepare for pull request. The original `allow` is very short and simple, but the `effect` is highly complex, which is why the decision ends up being spammed by the ai and does not stop being active until the `effect` can do something. The full `allow` is quoted in [the second post of this issue](https://github.com/moretrim/PFH/issues/39#issuecomment-524567696), but the `effect` is so large that it does not make sense to quote it all in one place. Instead I break it up to fully illustrate the problem.

The `potential` is here, and is occasionally good to consult to find flags that were created but mistakenly not used:
https://github.com/moretrim/PFH/blob/700259d9c193b4c32d6bfd82970f79f22d7c65e5/PFH/decisions/Italy.txt#L600-L671

- The part covering **Venice** is relatively easy to fix, because the `effect` is also simple. It makes sense to retain the flag, so that Italy cannot reclaim the province after dismantlement.

`allow`:
https://github.com/moretrim/PFH/blob/700259d9c193b4c32d6bfd82970f79f22d7c65e5/PFH/decisions/Italy.txt#L677

`effect`:
https://github.com/moretrim/PFH/blob/700259d9c193b4c32d6bfd82970f79f22d7c65e5/PFH/decisions/Italy.txt#L686-L693

- **Trient ( = South Tirol) and Trieste ( = Triest)** are covered in the same section, with more complex requirements in the `effect`. Please note the flag `italian_claimed_venetia` is set at the bottom but this likely needs to be changed to the flag that is referenced by the potential but not set anywhere: `italian_claimed_south_tirol` otherwise nothing will set this flag.

`allow`:
https://github.com/moretrim/PFH/blob/700259d9c193b4c32d6bfd82970f79f22d7c65e5/PFH/decisions/Italy.txt#L678-L679

`effect`:
https://github.com/moretrim/PFH/blob/700259d9c193b4c32d6bfd82970f79f22d7c65e5/PFH/decisions/Italy.txt#L695-L722

- **Savoia** is even more complex. Here, I elected to simply move requirements from the `effect` to the `allow`, rather than subject each one of them to too much scrutiny. Again, a flag from the `potential` was not properly set in the `effect`, `italian_reclaim_savoia` otherwise nothing will set this flag.

`allow`:
https://github.com/moretrim/PFH/blob/700259d9c193b4c32d6bfd82970f79f22d7c65e5/PFH/decisions/Italy.txt#L681

`effect`:
https://github.com/moretrim/PFH/blob/700259d9c193b4c32d6bfd82970f79f22d7c65e5/PFH/decisions/Italy.txt#L724-L748

- **Aosta** is not as complex as the above. Yet again, a flag from the `potential` needs to be added to the `effect`, `italian_claimed_aosta` otherwise nothing will set this flag.

`allow`: 
https://github.com/moretrim/PFH/blob/700259d9c193b4c32d6bfd82970f79f22d7c65e5/PFH/decisions/Italy.txt#L680

`effect`:
https://github.com/moretrim/PFH/blob/700259d9c193b4c32d6bfd82970f79f22d7c65e5/PFH/decisions/Italy.txt#L750-L763

**Comment by rogerburks** [comment]
Here is what the problem looks like, letting the game run for 8 days with Italy set to display a notice when using a decision:

![image](https://user-images.githubusercontent.com/17787203/76722131-945c6c00-66ff-11ea-9f95-5a4110a23dba.png)

**Comment by rogerburks** [comment]
First test with the new setup seems to be working:

![image](https://user-images.githubusercontent.com/17787203/76724539-83642880-6708-11ea-9fc3-6193e2d62dab.png)

**Comment by rogerburks** [comment]
Triest & Trient test successful:

![image](https://user-images.githubusercontent.com/17787203/76730292-2a9c8c00-6718-11ea-8ce8-b943e9c0aa50.png)

**Comment by rogerburks** [comment]
Player version of Savoia worked--even though there are 4 different ways for the ai to do this, the test was an unusually peaceful playthrough--will need to test again for ai.

![image](https://user-images.githubusercontent.com/17787203/76889382-bad6f000-6842-11ea-99ea-49b8e8a1d820.png)

**Comment by rogerburks** [comment]
Creative problem-solving: I convinced Germany to help assist in testing, and it seems that Savoia is probably working fine for the ai:

![image](https://user-images.githubusercontent.com/17787203/76891956-186d3b80-6847-11ea-8d90-94444109da60.png)

**Comment by rogerburks** [comment]
Seems like a good idea to itemize the Savoia requirements, to make sure they are working:

NOT = { has_global_flag = plombieres } 

This flag is set by the `plombieres` decision, which is used by Sardinia, before the existence of Italy, to ask France to assist in a war against Austria for Italian independence. The only way that this global flag would not exist, is if the `plombieres` decision is never used by Sardinia. This seems to be due to no deal existing between Sardinia and France to trade Savoy for help versus Austria.

has_global_flag = the_treaty_of_milan 

This flag is set by the decision `the_treaty_of_milan`, by France when Italy exists, to remove Italian cores from Aosta and Savoy (Annecy, Sciamberi [Chambéry], and Nizza [Nice]) in exchange for any other Italian cores that France may possess. Therefore, in this case `italia_irredenta` again adds Italian cores to Aosta and Annecy.

has_global_flag = france_dismantled 
`has_country_flag = in_great_war`

These last two flags seem self-explanatory: either France is dismantled or Italy is in a great war.

**Comment by rogerburks** [comment]
**Limitations of multi-job decisions**

This is a post that may be better placed in a more visible section, but I also want to use it as a brainstorming box before then. While there is a strong desire to make short and easily readable tooltips, the `allow` is not always able to achieve this goal, because it already has a different and more concrete role as a gatekeeper of decision activation.

The `potential` is only a gatekeeper of when a decision is visible. The `potential` can be satisfied for an entire play-through, the conditions of which may never satisfy requirements of the `allow`. Likewise, an `allow` can be valid for the entire play-through, but not possible to activate because the `potential` is never satisfied. For simple decisions, the `potential` can be complex while keeping the `allow` simple, such that a decision is not visible until the time when a decision is meant to be activated.

For complex decisions such as this one, which actually governs 4 different situations, any one of these 4 situations can make the decision visible. With a simple `allow`, the decision will be clickable when it is meant to be visible but not yet ready to yield its effect. This is the root cause of the spamminess of this decision. A simple `allow` existed because of good intentions of readability, but given the 4 different roles of the decision, the `allow` could not always satisfy its role as the final gatekeeper of decision activation.

In other words, there were unintended consequences resulting from not owning some of the provinces that were meant to be owned before those that were historically claimed later. Additionally, there was some difficulty with code in parts such as the section governing Savoy/Savoia.

Three possibilities exist to fix this: 

1. Make the `potential` exactly identical with the `allow`, such that the decision is not visible until exactly when it can do things. This would make the decision invisible for most of the game, such that it would surprise the user by its existence.

2. Split the decision to make it simpler, but thereby contribute to decision list bloat.

3. Trigger this through events instead, which lack some of the advantages of decisions. To me, a decision is something that is mean to be planned for, during the strategy of the play-through. An event can differ by having a possibility of never occurring due to variable circumstances, such as a `mean_time_to_happen` with modifiers. Events also have the possibility of permitting no choice, as opposed to decisions, which always permit a choice. I suspect that `italia_irredenta` would be better served as a set of events, although there is some cost in time, spent writing 4 new events with their own localisations.
